### ISSUE TITLE
Inappropriate way of sending locations is used

### ISSUE THREAD

**Issue by fynngodau** [signalapp/Signal-Android#13006]
The Signal app currently uses the following method of sharing a user's location, after the user has selected a location on the in-app map.

1. Load Google Map.
2. Wait for Google Map to finish loading.
3. Take snapshot of Google Map.
4. Remove Google Map and replace it with its snapshot.
5. Once the user hits send, send this snapshot as an image.

This is implemented in the following snippet:

https://github.com/signalapp/Signal-Android/blob/55af6ca84e32683662a96f9f621728d47fabae9e/app/src/main/java/org/thoughtcrime/securesms/components/location/SignalMapView.java#L77-L96

The two issues are:

* 1.-4. is an inappropriate way of generating a snapshot. It is better to [use a lite mode map instead](https://developers.google.com/maps/documentation/android-sdk/lite), because this implements exactly what is needed (non-interactive snapshot of a map).
    * An instance in which the current implementation is inappropriate is when `onMapLoaded` and `onSnapshotReady` have not yet been called, but the activity hosting the map is destroyed. In this case, the activity's `onDestroy` method must call the map's `onDestroy` method (which is not implemented). [Lite mode maps do not expect lifecycle methods to be called, except for `onCreate`.](https://developers.google.com/maps/documentation/android-sdk/lite#life_cycle_events)
    * In fact, Signal currently forgets to call a few of the lifecycle methods. See https://github.com/signalapp/Signal-Android/pull/13008.
    * Additional info: this is also related to https://github.com/signalapp/Signal-Android/issues/12723#issuecomment-1595810834, because microG does not implement `GoogleMap.OnMapLoadedCallback` correctly (https://github.com/microg/GmsCore/issues/1960).
* &#8288;5. is in violation of the [contract of the `snapshot` API](https://developers.google.com/maps/documentation/android-sdk/reference/com/google/android/libraries/maps/GoogleMap#public-void-snapshot-googlemap.snapshotreadycallback-callback), and by extension likely also of the terms of usage of the Google Maps API.
    > **Note:** Images of the map must not be transmitted to your servers, or otherwise used outside of the application. If you need to send a map to another application or user, send data that allows them to reconstruct the map for the new user instead of a snapshot.

**Comment by fynngodau** [comment]
Still current, don't close.

**Comment by ggdupont** [comment]
Just discovered the feature today (potentially a very useful feature). Then I understood it uses google map... to transmit a screenshot??

I share the technical concerns expressed by @fynngodau here: why send a screenshot when coordinates would be enough? On which I would add a privacy concern: would it be possible to use alternative privacy-friendly apps like OSM?

**Comment by fynngodau** [comment]
Still current, don't close.

**Comment by fynngodau** [comment]
Still current, don't close.

**Comment by fynngodau** [comment]
@stale You have been a bad bot.

**Comment by fynngodau** [comment]
Next: #13375.

**Comment by fynngodau** [comment]
Still current, don't close.

**Comment by ale5000-git** [comment]
The change in microG has been merged by mar-v-in.
Are there still problems with Signal?

**Comment by fynngodau** [comment]
@ale5000-git The Signal issue that was resolved with https://github.com/microg/GmsCore/pull/2182 was https://github.com/signalapp/Signal-Android/issues/12723#issuecomment-1595810834, which has since been closed by stalebot. This issue has little to do with that and is not specific to microG in itself.

**Comment by fynngodau** [comment]
Still current, don't close.

**Comment by dylanbabel** [comment]
Is there not a way to implement the use of other map services on a device? For example osmand or organic maps? Or simply accessing the GPS location of the device and sending the coordinates instead with links for google and apple maps and possibly others? I use iHunter for mapping offline in hunting areas and it shares locations like this.

**Comment by mxa** [comment]
Signal should simply get the coordinates from GPS and send them via a geo: link. The user receiving that link can then decide which app to use with the geo: intent. No need for Google Maps.

**Comment by Phoenix616** [comment]
It's really weird how this is implemented. Why send a link to google maps if the OS already has an intend for geo location?

I know it's completely unrelated from the obvious massive privacy and therefore security concern but like what if one wants to use the location in a different app that is not a map? (E.g. Öffi) Right now you are forced to manually copy the address/coordinates, that's just bad usability.

**Comment by mxa** [comment]
Bad bot! That the signal team seem to have different priorities hopefully doesn't mean this will never be fixed.

**Comment by MichaelZ4714** [comment]
I really wish that this problem will be solved soon!!

**Comment by c4tb3rt** [comment]
> Signal should simply get the coordinates from GPS and send them via a `geo:` link. The user receiving that link can then decide which app to use with the `geo:` intent. No need for Google Maps at all.

This really seems like the best idea.

**Comment by jayaddison** [comment]
> Signal should simply get the coordinates from GPS and send them via a `geo:` link. The user receiving that link can then decide which app to use with the `geo:` intent. No need for Google Maps at all.

On the other hand: it is helpful for sighted users to have visual imagery of a map location; a standalone `geo:` URI alone might not provide that.

**Comment by mxa** [comment]
> On the other hand: it is helpful for sighted users to have visual imagery of a map location; a standalone geo: URI alone might not provide that.

True, but since sharing the sent location with a map provider is a privacy issue, the preview feature should at best be a n opt-in preference, not the default.

**Comment by jayaddison** [comment]
> > On the other hand: it is helpful for sighted users to have visual imagery of a map location; a standalone geo: URI alone might not provide that.
> 
> True, but since sharing the sent location with a map provider is a privacy issue, the preview feature should at best be a n opt-in preference, not the default.

After spending some time thinking about it, it makes sense to me that the app has to connect out to retrieve a map image from a remote service, because that data -- that evolves on a continuous basis, for example businesses, road layouts, etc -- is not going to ship inside the app with today's mobile network / map data size tradeoffs.

Also, it could be an option for every recipient to make a decision about whether to display the sender's location as a map preview.  However, given that Signal supports large-ish group sizes, a single sender causing `n` (where `n` could easily be greater than 10) recipients to reveal their location to a remote map service, means we should probably prefer for the sender -- precisely 1 user -- to choose.

I do think your suggestion that it could be opt-in is reasonable.  I would suggest perhaps a prompt-on-first-share, with a choice of user selections from:

- Always allow map previews when I share my location.
- Allow map previews when I share my location within this conversation.
- Ask whether to provide a map preview every time I share my location.
- Only send `geo:` URIs when I share my location.

An orthogonal issue is: what map preview providers are supported (currently, AFAICT, this is Google Maps only -- however others could be added).

**Comment by clijsters** [comment]
Just jumping in with a user-perspective: I am using /e/OS, which has microG enabled - the map preview works without the classic google maps invocation. But the link which gets generated is a google maps link.

What my recipient gets, is a map picture that suggests its a MapLibre location (no one know the difference between map providers and navigation apps) and a link to `maps.google.com`

I'd say a `geo:` link as default would be a good start and a user selection for if a preview should get generated, like suggested by @jayaddison makes sense to me. 

I'm impressed, that this topic is open since 2023

**Comment by hamishmb** [comment]
This feature doesn't work at all for me, as I am on LineageOS without Google Play Services, which is disappointing.

**Comment by dylanbabel** [comment]
Many privacy focused users like the encryption of signal but do not have Google
Play Services installed making. It impossible to easily send a pin. 


Dylan Barlowe Abel BBA - Accounting

This Email is intended only for the person to whom it is addressed and may conta
in confidential and
privileged material. Any review, retransmission, dissemination or other use that
 a person other than the addressee makes of this communication is prohibited and
 any reliance or decisions made based on it, are the responsibility of such pers
on. If you received this in error, please contact the sender and destroy all cop
ies of this Email.


Sent from Proton Mail for Android.


-------- Original Message --------

**Comment by bordenc** [comment]
Would this be related to the fact that Signal recognises geo URIs as phone numbers and not geo URIs? I second the others that the users should decide what mapping app to use, not Signal.

**Comment by dylanbabel** [comment]
Yes.

What it comes down to is a user who does not have Google Play Services should
still be able to send their location easily as the signal app can just query the
device for coordinates and append those onto any link the user selects for
mapping apps. 


Dylan Barlowe Abel BBA - Accounting

This Email is intended only for the person to whom it is addressed and may conta
in confidential and
privileged material. Any review, retransmission, dissemination or other use that
 a person other than the addressee makes of this communication is prohibited and
 any reliance or decisions made based on it, are the responsibility of such pers
on. If you received this in error, please contact the sender and destroy all cop
ies of this Email.


Sent from Proton Mail for Android.



-------- Original Message --------
